app/src/common/shared/com/igalia/wolvic/ui/adapters/ThumbnailAsyncTask.java

```diff
@@ -0,0 +1,143 @@
+    }
+
+    private static final Size DEFAULT_SIZE = new Size(96, 96);
+    private static final int DEFAULT_SIZE_KIND = MediaStore.Images.Thumbnails.MICRO_KIND;
+
+    private final ContentResolver mContentResolver;
+    private final Uri mFileUri;
+    private final OnSuccessDelegate mOnSuccessDelegate;
+    private CancellationSignal mCancellationSignal;
+
+    public ThumbnailAsyncTask(@NonNull Context context, @NonNull Uri fileUri, OnSuccessDelegate onSuccessDelegate) {
+
+        Log.e(LOGTAG, "new ThumbnailAsyncTask");
+
+        mContentResolver = context.getContentResolver();
+        mFileUri = fileUri;
+        mOnSuccessDelegate = onSuccessDelegate;
+    }
+
+    @Override
+    protected Bitmap doInBackground(Void... voids) {
+        Log.e(LOGTAG, "doInBackground  mFileUri = " + mFileUri);
+
+        if (UrlUtils.isFileUri(mFileUri.toString())) {
+            File file = new File(mFileUri.getPath());
+            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
+                return createFileThumbnail(file);
+            } else {
+                return createFileThumbnailLegacy(file);
+            }
+        } else {
+            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
+                return createContentThumbnail(mFileUri);
+            } else {
+                return createContentThumbnailLegacy(mFileUri);
+            }
+        }
+    }
+
+    @Override
+    protected void onCancelled() {
+        super.onCancelled();
+
+        if (mCancellationSignal != null) {
+            mCancellationSignal.cancel();
+            mCancellationSignal = null;
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.Q)
+    private Bitmap createFileThumbnail(@NonNull File file) {
+        Log.e(LOGTAG, "createFileThumbnail file = " + file);
+        String extension = MimeTypeMap.getFileExtensionFromUrl(file.getPath());
+        Log.e(LOGTAG, "  extension = " + extension);
+        if (extension == null) {
+            return null;
+        }
+        String mimeFromExtension = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
+        mCancellationSignal = new CancellationSignal();
+        Log.e(LOGTAG, "  mimeFromExtension = " + mimeFromExtension);
+
+        try {
+            if (mimeFromExtension.startsWith("audio")) {
+                Log.e(LOGTAG, "  createAudioThumbnail");
+                return ThumbnailUtils.createAudioThumbnail(file, DEFAULT_SIZE, mCancellationSignal);
+            } else if (mimeFromExtension.startsWith("video")) {
+                Log.e(LOGTAG, "  createVideoThumbnail");
+                return ThumbnailUtils.createVideoThumbnail(file, DEFAULT_SIZE, mCancellationSignal);
+            } else if (mimeFromExtension.startsWith("image")) {
+                Log.e(LOGTAG, "  createImageThumbnail");
+                return ThumbnailUtils.createImageThumbnail(file, DEFAULT_SIZE, mCancellationSignal);
+            }
```

Comment:
Unftorunate that there isn't a thumbnailutils method that takes the mime type and calls the appropriate method

Reply:
Yes, that's true.